doc/design/zarr-support-2.md 

```diff
@@ -0,0 +1,94 @@
+# Abstract
+
+Zarr files are stored in their entirety in S3.
+Zarr files are represented by a single Asset through the API.
+A treehash is computed as the zarr file is uploaded.
+Each node in the treehash in the hash is stored in S3 to avoid bloating the DB.
+When published, the zarr file is copied to a separate location in the bucket so that changes to the draft cannot affect the published file.
+
+# Requirements
+
+1. Zarr files are stored in a "directory" in S3.
+1. Each zarr file corresponds to a single Asset.
+1. The CLI uses some kind of tree hashing scheme to compute a checksum for the entire zarr file.
+1. The API verifies the checksum _immediately_ after upload.
+1. The system can theoretically handle zarr files with ~1 million subfiles, each of size 64 * 64 * 64 bytes ~= 262 kilobytes.
+1. Zarr metadata must be mutable for zarr files in draft dandisets.
+
+# Implementation
+
+## API endpoints
+
+* POST /api/zarr_files/
+  Start a new zarr file.
+  Returns a zarr ID
+* GET /api/zarr_files/{zarr_id}/
+  Returns some information about the zarr file, like name, S3 path+url, and checksum
+* POST /api/zarr_files/{zarr_id}/batch/start/
+  Ask to upload a batch of subfiles as part of the zarr file.
+  No more batch uploads are permitted until this one is completed or aborted.
```

Comment:
agree -- no simultaneous batches, but allow different users to upload (whenever there is no ongoing batch) to the same "zarr file".

The question remains -- how will we ensure the consistency / integrity of the entire "zarr file" and not just a particular batch?  (in particular in the aforementioned "multiple users can upload", but can manifest in the same user upload while local "zarr file" gets changed between batches) .  I guess it could be up to a CLI to get a full "checksum" for a "zarr file" upon completion and compare with the local version (which might require heavy traversal, but may be fscache would be of help)?